### PR TITLE
Added HealthCommand, based off the existing FreezeCommand for modifyi…

### PR DESCRIPTION
**Pull request by Melodi17**
…ng living entities' health.

This is a useful command in my opinion as it adds an easy way to modify any entities' health even players.

### CODE REVIEW

**Comment by TheDeathlyCow** [comment]
Thanks for this! I will try and get to it later today.

**Comment by Melodi17** [comment]
Sorry, wrong button

File: src/main/java/com/github/thedeathlycow/tdcdata/server/command/HealthCommand.java [hunk 1]
```diff
@@ -0,0 +1,281 @@
...
+     * @param source      The source of the command
+     * @param targets     The collection of {@link Entity}s affected by this command
+     * @param amount      The of health to set on each target.
+     * @param shouldClamp Whether the amount should be clamped
+     * @return Returns the sum of the healths of each target after execution
+     */
+    private static int set(final ServerCommandSource source, final Collection<? extends Entity> targets, final int amount, final boolean shouldClamp) {
+        int sum = 0;
+        for (Entity target : targets) {
+            if (!(target instanceof LivingEntity livingTarget)) {
```
Comment (TheDeathlyCow) [inline src/main/java/com/github/thedeathlycow/tdcdata/server/command/HealthCommand.java]: Either all `LivingEntity`s selected by this command should be affected by its execution, or no entities should be affected. It should not return early. Also, the message should only show if a single entity is selected.

File: src/main/java/com/github/thedeathlycow/tdcdata/server/command/HealthCommand.java [hunk 2]
```diff
@@ -0,0 +1,281 @@
...
+     * @param targets     The targeted {@link LivingEntity}s to adjust the health of.
+     * @param amount      The amount to adjust by
+     * @param shouldClamp Whether to clamp the final amount. Defaults to true.
+     * @param isRemoving  Whether this command is removing or adding.
+     * @return Returns the sum of the healths of each target after execution
+     */
+    private static int adjust(final ServerCommandSource source, final Collection<? extends Entity> targets, final int amount, final boolean shouldClamp, final boolean isRemoving) {
+        int sum = 0;
+        for (Entity target : targets) {
+            if (!(target instanceof LivingEntity livingTarget)) {
```
Comment (TheDeathlyCow) [inline src/main/java/com/github/thedeathlycow/tdcdata/server/command/HealthCommand.java]: Same as previous comment.

File: src/main/java/com/github/thedeathlycow/tdcdata/server/command/HealthCommand.java [hunk 3]
```diff
@@ -0,0 +1,281 @@
...
+     * @return Returns the amount of health the target has.
+     */
+    private static int getCurrent(final ServerCommandSource source, final Entity target) {
+        if (!(target instanceof LivingEntity)) {
+            Text msg = Text.literal(String.format(GET_CURRENT_FAIL, target.getDisplayName().getString()));
+            source.sendError(msg);
+            return -1;
+        }
+
+        int amount = Math.round(((LivingEntity) target).getHealth());
```
Comment (TheDeathlyCow) [inline src/main/java/com/github/thedeathlycow/tdcdata/server/command/HealthCommand.java]: I'm not sure `Math.round` is the best function to use here; perhaps `MathHelper.floor` should be used instead? Reason being that if you have 19.5 health, your health bar will render 19 health points. However, `Math.round` would round 19.5 to 20.

**Comment by Melodi17** [comment]
[Health-Command.md](https://github.com/TheDeathlyCow/tdc-data/files/9075350/Health-Command.md)
Here is a wiki page I wrote for it (based on the freeze page), if that helps.

File: src/main/java/com/github/thedeathlycow/tdcdata/server/command/HealthCommand.java (hunk 1 above)
Comment (Melodi17) [reply src/main/java/com/github/thedeathlycow/tdcdata/server/command/HealthCommand.java]: Fixed, the code now checks the targets beforehand

File: src/main/java/com/github/thedeathlycow/tdcdata/server/command/HealthCommand.java (hunk 2 above)
Comment (Melodi17) [reply src/main/java/com/github/thedeathlycow/tdcdata/server/command/HealthCommand.java]: Fixed, same as previous comment

File: src/main/java/com/github/thedeathlycow/tdcdata/server/command/HealthCommand.java (hunk 3 above)
Comment (Melodi17) [reply src/main/java/com/github/thedeathlycow/tdcdata/server/command/HealthCommand.java]: Addressed this, changed `Math.round` to `MathHelper.floor`

**Comment by TheDeathlyCow** [comment]
I've been thinking some more about how this command could handle living vs non living entities, and I think perhaps the best thing to do might be to make the target of the command a single entity for all cases. The vanilla command [/attribute](https://minecraft.fandom.com/wiki/Commands/attribute) is a somewhat similar idea to this, and it only allows a single entity to be targeted at a time. You can of course use `/execute as` to target multiple entities, and any non living targets will effectively be skipped. It is a little more laggy than targeting multiple entities, but still much better than `/data` and in my opinion more consistent with vanilla.

**Comment by Melodi17** [comment]
I like this idea, error handling/messages will be less complex and it will seem to be a more "vanilla" feature. For consistency, should we do something similar with the `/freeze` command? I will start implementing your idea as soon as I can (5-9 hours from this message).

**Comment by TheDeathlyCow** [comment]
The reason `/freeze` can still target any entity is because, unlike health, freezing is something any `Entity` can do, it is not limited to a certain subclass of `Entity`.

**Review by TheDeathlyCow — APPROVED** [review]

**Review by TheDeathlyCow — APPROVED** [review]

**Pull request merged**